Allow non-static Cartesian factory methods with PER_CLASS lifecycle

robtimus opened the pull request:
```
Allow non-static Cartesian factory methods with PER_CLASS (#628)

JUnit allows non-static factory methods as long as the test class is
annotated with `@Testinstance(Lifecycle.PER_CLASS)`. The same is
currently not true for `@CartesianTest.MethodFactory`.
This commit fixes that.

The existing non-static test is unmodified, indicating that
non-static is still not allowed if the test class is not annotated
with `@Testinstance(Lifecycle.PER_CLASS)`.

PR: #628
```

---
**PR checklist**

Documentation (general)
* [ ] There is documentation (Javadoc and site documentation; added or updated)
* [ ] Only one sentence per line (especially in `.adoc` files)
* [ ] Javadoc uses formal style, while sites documentation may use informal style

Code
* [X] Code adheres to code style, naming conventions etc.
* [X] Successful tests cover all changes
* [X] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [X] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [X] A prepared commit message exists
* [X] The list of contributions inside `README.md` mentions the new contribution (real name optional)

Michael1993 reviewed (changes requested):
Hi! 👋 
Thank you for this, looks great so far! Could you please also add an example with a non-static factory method to the documentation? Either to showcase the wrong usage (i.e. with a missing `@TestInstance` annotation) or the correct usage.

Michael1993 commented:
Don't forget spotless! 👿

rob-spoor commented:
> Don't forget spotless! 👿

The errors are an update in the year in the licenses. I think that the spotless configuration needs to be updated, because the second year should be 2022, not 2021. Spotless suggest changing it back to 2021.

Michael1993 commented:
Please rebase or merge main.

Michael1993 commented:
> > Don't forget spotless! 👿
> 
> The errors are an update in the year in the licenses. I think that the spotless configuration needs to be updated, because the second year should be 2022, not 2021. Spotless suggest changing it back to 2021.

Whoops. 😅

robtimus commented:
The rebase is complete, I'll work on the documentation shortly.

robtimus commented:
I've added the requested example.

While working on the example I found a case where my change would have caused failures - using a non-static factory method from another class. I've improved the check that determines whether or not the factory method must be required, and added some tests. There are now three, in addition to the unmodified tests that were already on `main`:
* Using a non-static factory method from another class in a class annotated with `@TestInstance(PER_CLASS)` still fails.
* Using a static factory method from another class in a class annotated with `@TestInstance(PER_CLASS)` still works.
* Using a non-static factory method from the same class in a class annotated with `@TestInstance(PER_CLASS)` works (I already added this test before).

As a bonus I added support for nested test classes in `PioneerTestKit`. I made those changes generic enough to also support entire nested test classes and no-argument methods in nested test classes.

Michael1993 reviewed (approved):
LGTM 👍 

Thank you!

nipafx commented on docs/cartesian-product.adoc:
```diff
@@ -387,23 +387,46 @@ class MyCartesianTestClass {
+class MyCartesianTestClass {
+	@Nested
+	@TestInstance(Lifecycle.PER_CLASS)
+	MyNestedCartesianTestClass {
```
```suggestion
	class MyNestedCartesianTestClass {
```

Right?

robtimus replied:
Right.

I did a rebase on main, and extracted the example to `CartesianTestExtensionDemo`, so these errors should now be caught..

nipafx commented on docs/cartesian-product.adoc:
```diff
@@ -387,23 +387,46 @@ class MyCartesianTestClass {
+			// passing test code
+		}
+
+		ArgumentSets provideArguments() {
```
```suggestion
		// this provider method doesn't have to be static
		ArgumentSets provideArguments() {
```

robtimus replied:
In fact, it can't even be static because it's in a non-static nested class. I've added a comment about that, as well as why the `@TestInstance` annotation is necessary.

nipafx reviewed (commented):
I started a review a week ago and forgot to send it, stupid me. Just added another comment - can you take a look at this?

robtimus commented on docs/cartesian-product.adoc:
```diff
@@ -387,23 +387,46 @@ class MyCartesianTestClass {
+There are multiple requirements the factory method has to fulfill to qualify:
 
 - must have the same name as the test method (or its name must be specified via the `factory` attribute)
-- must be `static`
+- must be `static` (unless the test class is annotated with `@TestInstance(Lifecycle.PER_CLASS)` and the factory method is defined in the test class)
 - must have **no** parameters
 - must return `ArgumentSets`
 - must register values for every parameter exactly once
 - must register values in order
```
Didn't I say that:

>  and the factory method is defined in the test class

nipafx replied:
You did. 🤦🏾

robtimus commented:
@nipafx can you check again? I've rebased on main and split off the example, as was recently done on main for other Cartesian test examples.

nipafx commented on src/demo/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtensionDemo.java:
```diff
@@ -279,4 +282,30 @@ void testWithCustomDisplayName(@Values(strings = { "0", "1" }) String a,
+				// passing test code
+			}
+
+			// this provider method cannot be static because it is in a non-static nested class
```
Since Java 16, this is not longer correct. To make records (which are always static) easier to use in non-static nested classes, this limitation was lifted and non-static nested classes can now have static members. See [this compiling example](https://github.com/nipafx/demo-java-x/blob/main/src/main/java/dev/nipafx/demo/java16/lang/staticinner/OuterClass.java).

I think it's enough to point out that the method doesn't have to be static, but if you want, you can also add that it couldn't even be until Java 16.

robtimus replied:
I wasn't aware of this. I've used your original suggestion.

nipafx commented on README.md:
```diff
@@ -119,6 +119,7 @@ The least we can do is to thank them and list some of their accomplishments here
 #### 2022
 
 * [Filip Hrisafov](https://github.com/filiphr) contributed the [JSON Argument Source](https://junit-pioneer.org/docs/json-argument-source/) support (#101 / #492)
+* [Rob Spoor](https://github.com/robtimus) made it possible to use non-static factory methods for `@CartesianTest.MethodFactory`, as long as the test class is annotated with `@TestInstance(Lifecycle.PER_CLASS)` (#628)
```
```suggestion
* [Rob Spoor](https://github.com/robtimus) enabled non-static factory methods for `@CartesianTest.MethodFactory` (#628)
```

No reason to go into detail here.

robtimus replied:
Updated

nipafx commented:
Merged - thank you for your contribution @robtimus! :heart:

robtimus commented:
You're welcome 😄
